src/shims/sync.rs

```diff
@@ -454,135 +456,325 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     fn pthread_mutex_destroy(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        if mutex_get_locked_count(this, mutex_op)?.to_u32()? != 0 {
+        let id = mutex_get_or_create_id(this, mutex_op)?;
+
+        if this.mutex_is_locked(id) {
             throw_ub_format!("destroyed a locked mutex");
         }
 
         mutex_set_kind(this, mutex_op, ScalarMaybeUndef::Undef)?;
-        mutex_set_locked_count(this, mutex_op, ScalarMaybeUndef::Undef)?;
-        mutex_set_blockset(this, mutex_op, ScalarMaybeUndef::Undef)?;
+        mutex_set_id(this, mutex_op, ScalarMaybeUndef::Undef)?;
 
         Ok(0)
     }
 
     fn pthread_rwlock_rdlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
-        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+        let active_thread = this.get_active_thread()?;
 
-        if writers != 0 {
-            // The lock is locked by a writer.
-            assert_eq!(writers, 1);
-            let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
-            this.block_active_thread(reader_blockset)?;
+        if this.rwlock_is_write_locked(id) {
+            this.rwlock_enqueue_reader(id, active_thread);
+            this.block_thread(active_thread)?;
```

Comment:
Similar here, is there ever a case where we want to enqueue a reader and *not* block the thread? If no, maybe `rwlock_enqueue_reader` should take care of that?